atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
-            for sid in host.free_slots:
-                if host.lock(sid, prid):
-                    return host, sid
-        raise Exception('No availiable host found')
-
-
-if __name__ == '__main__':
-    my_hosts = [Host('username', 'hostname')]
-    mngr = ResourceManager(my_hosts)
-    # locking
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            stdin, stdout, stderr = client.exec_command(
+                cmd, timeout=SSH_COMMAND_TIMEOUT)
+        except paramiko.ssh_exception.AuthenticationException as ex:
+            logger.warning("SSH authentication failed. %s", ex) #  TODO: Question - Probably not needed for exec command
+        except paramiko.ssh_exception.ChannelException as ex:
+            logger.warning("Attempt to open a new Channel failed. %s", ex) #  TODO: Question - Probably not needed for exec command
+        except paramiko.ssh_exception.NoValidConnectionsError as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return stdout, stderr, client.recv_exit_status()
+
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
+        stdin, stdout, stderr = self._ssh_run_remote_cmd(client, cmd)
+        try:
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON  #  TODO: Question - Do we want to fail build on loading invalid json from remote host????
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
+        except TypeError as ex:  # Unable to decode the bytes
+            logger.warning("Json load failed. Unable to decode the bytes. %s", ex)
+            return None
+        return json_stdout
+
+    def available_slots(self) -> list[int]:
+        """
+        Returns list of available slots
+
+        Slots should be randomized here, for smoother locking, because there will be
+        hunger games when every pipeline will try to lock slot with lowest number first
+        """
+        logger.debug("Retrieve list of available slots on host %s", self._hostname)
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            available_slots = []
+            for slot_id in range(resource_info.get('max_slot_count', 1)):
+                cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+                slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+                if slot is None:  # Json load failed - file on remote host might be corrupted
```

Comment:
How are slots created initially?

Reply:
I believe via our configuration playbook for remote hosts, but I'll check it in design.

Reply:
Added note to ticket for follow up PR.